### ISSUE TITLE
Reading and writing part 2: files and databases

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8463]
Reading and writing, part 2: files and databases<br>
<br>
https://ift.tt/FwK1zNx<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<div>
<p>This article is part of a five-part series on race conditions. The complete series is:</p>
<p>&ndash;</p>
<p>In <a href="https://blog.jcoglan.com/2020/09/28/reading-and-writing-part-1/">part 1</a> we introduced the <code>Location</code> type, which provides operations for reading and writing a value, and saw how it occurs inside programs as they manage the data they hold in memory. This led us towards some strategies for dealing with the safety problems this type creates in concurrent code, including the use of object orientation and mutexes to manage access to data. We will now apply this concept to data stored outside program memory &ndash; on disk, in databases, and other external services.</p>
<p>Let&rsquo;s start with maybe the simplest form of external storage: a single file. We&rsquo;ll modify our counter example to store the counter&rsquo;s value in a file named <code>counter.txt</code>, whose initial content is <code>0</code>.</p>
<div>
<div>
<pre><code><span>filename</span> <span>=</span> <span>"counter.txt"</span>

<span>threads</span> <span>=</span> <span>(</span><span>1</span><span>..</span><span>4</span><span>).</span><span>map</span> <span>do</span>
  <span>Thread</span><span>.</span><span>new</span> <span>do</span>
    <span>100</span><span>.</span><span>times</span> <span>do</span>
      <span>tmp</span> <span>=</span> <span>File</span><span>.</span><span>read</span><span>(</span><span>filename</span><span>).</span><span>strip</span><span>.</span><span>to_i</span>      <span># read file</span>
      <span>content</span> <span>=</span> <span>"</span><span>#{</span><span>tmp</span> <span>+</span> <span>1</span><span>}</span><span>\n</span><span>"</span>
      <span>File</span><span>.</span><span>write</span><span>(</span><span>filename</span><span>,</span> <span>content</span><span>)</span>             <span># write file</span>
    <span>end</span>
  <span>end</span>
<span>end</span>

<span>threads</span><span>.</span><span>each</span><span>(</span><span>&amp;</span><span>:join</span><span>)</span>
<span>puts</span> <span>File</span><span>.</span><span>read</span><span>(</span><span>filename</span><span>)</span>
</code></pre>
</div>
</div>
<p>We have the same <code>Location</code> pattern here that we saw in previous examples: we read from the <code>counter.txt</code> file, generate some new value based on the data we read, and then write that value back to the same file. Ignoring the fact that file reads and writes to disk can fail, these are occurrences of the <code>read() -&gt; T</code> and <code>write(T)</code> operations. If you run this, you&rsquo;ll see that almost all the updates get lost &ndash; this happens because most of the <code>read</code> calls are producing the same value, as they don&rsquo;t wait for any <code>write</code> calls currently in progress to finish.</p>
<p>This is a really common operation in command-line programs, for example for editing a configuration file, or automatically formatting source code. If this operation is done concurrently, then one task can have its data overwritten by another that was running at the same time, giving us a <em>lost update</em> bug. To make sure every task sees the changes made by all others, the tasks must be made to run sequentially.</p>
<p>When dealing with data in memory, we wrapped the data we needed to protect in abstractions to control the program&rsquo;s access to it. When dealing with files on disk, we can&rsquo;t <em>wrap</em> data in the same way: every process running on the machine has the same access to the filesystem, and one process can&rsquo;t insert an abstraction that affects every other program&rsquo;s access. But the filesystem does offer some operations to help programs co-ordinate their use of files, and make concurrent file access safe.</p>
<p>The <a href="http://manpages.ubuntu.com/manpages/focal/en/man2/open.2.html"><code>open()</code> system call</a> accepts a set of flags that control whether the file is opened for reading or writing or both, whether to create the file if it doesn&rsquo;t exist, whether to truncate or append to it, and so on. By combining the flags <code>O_RDWR</code>, <code>O_CREAT</code> and <code>O_EXCL</code>, we can ask the kernel to open a file for reading and writing, and to create the file <em>only</em> if it doesn&rsquo;t exist; if the file already exists then <code>open()</code> will fail.</p>
<p>Let&rsquo;s rewrite the loop body of the above program to make use of this. Before trying to read <code>counter.txt</code>, we open <code>counter.txt.lock</code> with the above flags. This will only succeed if that file does not exist, so if we manage to open the file, that means nobody else can open it until we delete it again &ndash; the file acts like a mutex. If we fail to open the file, we retry until we&rsquo;re successful; this is like a mutex&rsquo;s <code>lock()</code> method blocking until the lock is acquired. Once the lockfile is open, we can read the <code>counter.txt</code> file to get the current value. To store the new value, we write to the <em>lockfile</em>, and then rename the lockfile to the name of the file we want to update, i.e. we rename <code>counter.txt.lock</code> to <code>counter.txt</code>. This achieves two things: first, because <a href="http://manpages.ubuntu.com/manpages/focal/en/man2/rename.2.html"><code>rename()</code></a> is atomic, it means <code>counter.txt</code> never exists in a partially complete state, where not all the new data has finished being written to it. Second, it effectively removes the lockfile so another task can claim it.</p>
<div>
<div>
<pre><code>      <span>begin</span>
        <span>flags</span> <span>=</span> <span>File</span><span>::</span><span>RDWR</span> <span>|</span> <span>File</span><span>::</span><span>CREAT</span> <span>|</span> <span>File</span><span>::</span><span>EXCL</span>
        <span>lock</span> <span>=</span> <span>File</span><span>.</span><span>open</span><span>(</span><span>"</span><span>#{</span><span>filename</span><span>}</span><span>.lock"</span><span>,</span> <span>flags</span><span>)</span>
      <span>rescue</span> <span>Errno</span><span>::</span><span>EEXIST</span>
        <span>retry</span>
      <span>end</span>
      <span>tmp</span> <span>=</span> <span>File</span><span>.</span><span>read</span><span>(</span><span>filename</span><span>).</span><span>strip</span><span>.</span><span>to_i</span>
      <span>lock</span><span>.</span><span>puts</span><span>(</span><span>"</span><span>#{</span><span>tmp</span> <span>+</span> <span>1</span><span>}</span><span>"</span><span>)</span>
      <span>lock</span><span>.</span><span>close</span>
      <span>File</span><span>.</span><span>rename</span><span>(</span><span>"</span><span>#{</span><span>filename</span><span>}</span><span>.lock"</span><span>,</span> <span>filename</span><span>)</span>
</code></pre>
</div>
</div>
<p>This is a much safer way to update a file, which avoids race conditions and partially-complete data. It&rsquo;s how Git updates its index, branch pointers and config files, to make sure that no two processes try to update the same file at the same time and end up overwriting each other&rsquo;s changes. In part 1, we introduced the <code>Mutex</code> type, which gates the data access operations behind a successful call to <code>lock()</code>:</p>
<div>
<div>
<pre><code><span>Mutex</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>=</span> <span>{</span>
    <span>lock</span><span>()</span> <span>-&gt;</span> <span>Lock</span><span>&lt;</span><span>T</span><span>&gt;</span><span>;</span>
<span>}</span>

<span>Lock</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>=</span> <span>{</span>
    <span>read</span><span>()</span> <span>-&gt;</span> <span>T</span><span>;</span>
    <span>write</span><span>(</span><span>T</span><span>);</span>
    <span>unlock</span><span>();</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>The <code>File</code> API gives us something very similar, but its <code>open()</code> method can fail rather than blocking indefinitely:</p>
<div>
<div>
<pre><code><span>FileSystem</span> <span>=</span> <span>{</span>
    <span>open</span><span>()</span> <span>-&gt;</span> <span>Option</span><span>&lt;</span><span>File</span><span>&gt;</span><span>;</span>
<span>}</span>

<span>File</span> <span>=</span> <span>{</span>
    <span>read</span><span>()</span> <span>-&gt;</span> <span>String</span><span>;</span>
    <span>write</span><span>(</span><span>String</span><span>);</span>
    <span>close</span><span>();</span>
<span>}</span>
</code></pre>
</div>
</div>
<p><code>open()</code> returns <code>Option&lt;File&gt;</code>, which means it returns a <code>File</code> on success, but may also return nothing when it fails. We can&rsquo;t call <code>read()</code> or <code>write()</code> until we have a <code>File</code>, and <code>open()</code> makes sure we can&rsquo;t get one until it&rsquo;s safe. Letting this operation fail gives us the choice of giving up rather than waiting indefinitely; we can regain <code>lock()</code> semantics by calling <code>open()</code> in a loop until it succeeds, as we did above.</p>
<p>So we have seen that the Unix file system provides tools for programs to use it concurrently and safely co-ordinate their activity &ndash; we don&rsquo;t need to rely on any special mechanism in our programming languages to manage file access. But while files are the bedrock of the Unix system, many applications don&rsquo;t deal with them directly. They instead deal with databases and other services that provide powerful ways to store, index and query their data. Here we see a rich variety of tools for handling concurrency safely, but it&rsquo;s also where we see a lot of concurrency errors occur in the wild.</p>
<p>The paper <a href="http://www.bailis.org/papers/feral-sigmod2015.pdf">Feral Concurrency Control</a> catalogs the problems with how <a href="https://rubyonrails.org/">Ruby on Rails</a> in particular manages data integrity. Many other frameworks have similar problems, but Rails is what I&rsquo;m most familiar with in the web application space and serves as a good example of how developers are discouraged from using the tools their database provides to handle concurrency. We&rsquo;ll use our counter example to demonstrate a few ways that a database can give us safe concurrency, before talking about how these problems tend to come up in web applications.</p>
<p>For these examples, I&rsquo;ll be using <a href="https://www.postgresql.org/">PostgreSQL</a> with the following schema and model: a <code>Counter</code> has a <code>name</code> and a <code>value</code>, which must not be null and defaults to <code>0</code>.</p>
<div>
<div>
<pre><code><span>ActiveRecord</span><span>::</span><span>Schema</span><span>.</span><span>define</span> <span>do</span>
  <span>create_table</span> <span>:counters</span><span>,</span> <span>force: </span><span>true</span> <span>do</span> <span>|</span><span>t</span><span>|</span>
    <span>t</span><span>.</span><span>string</span> <span>:name</span><span>,</span> <span>null: </span><span>false</span>
    <span>t</span><span>.</span><span>integer</span> <span>:value</span><span>,</span> <span>null: </span><span>false</span><span>,</span> <span>default: </span><span>0</span>
  <span>end</span>
<span>end</span>

<span>class</span> <span>Counter</span> <span>&lt;</span> <span>ActiveRecord</span><span>::</span><span>Base</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>As usual, we start up four threads, and have each of them attempt to increment the counter repeatedly. In our first implementation of <code>inc_counter</code>, we do this by calling <code>Counter.find_by</code> to read the counter record from the database, and then <code>Counter#update</code> to write a new <code>value</code>, based on the one we initially read.</p>
<div>
<div>
<pre><code><span>Counter</span><span>.</span><span>create</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>

<span>def</span> <span>inc_counter</span>
  <span>ctr</span> <span>=</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
  <span>ctr</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>ctr</span><span>.</span><span>value</span> <span>+</span> <span>1</span><span>)</span>
<span>end</span>

<span>threads</span> <span>=</span> <span>(</span><span>1</span><span>..</span><span>4</span><span>).</span><span>map</span> <span>do</span>
  <span>Thread</span><span>.</span><span>new</span> <span>do</span>
    <span>100</span><span>.</span><span>times</span> <span>{</span> <span>inc_counter</span> <span>}</span>
  <span>end</span>
<span>end</span>

<span>threads</span><span>.</span><span>each</span><span>(</span><span>&amp;</span><span>:join</span><span>)</span>
<span>p</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
</code></pre>
</div>
</div>
<p>As you can no doubt guess by now, this example does not work. On my machine, the counter ends up having a value somewhere between 100 and 200, so a majority of the updates this program makes are lost. We can see why by examining the SQL that is logged when we run this code:</p>
<div>
<div>
<pre><code><span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'my-counter'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>BEGIN</span><span>;</span>
<span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>We <code>SELECT</code> the record, loading a copy of it into memory in the <code>ctr</code> variable. Calling <code>ctr.update</code> triggers the <code>BEGIN</code>/<code>UPDATE</code>/<code>COMMIT</code> queries that update the record, writing the new <code>value</code>. Rails automatically wraps all updates in a transaction, so that any validations and life-cycle callbacks triggered by the update are run atomically with it. But, the initial <code>SELECT</code> is not inside this transaction and is done without holding a lock on the record. That means two concurrent tasks can both run the <code>SELECT</code> query, both get an identical copy of its current state, and then both <code>UPDATE</code> the record with the same new <code>value</code>.</p>
<div>
<div>
<pre><code><span>Task</span> <span>1</span>                          <span>|</span> <span>Task</span> <span>2</span>
<span>--------------------------------+--------------------------------</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>...</span>      <span>|</span>
    <span>--&gt; value = 1 |</span>
                                <span>|</span> <span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>...</span>
                                <span>|</span>     <span>--&gt; value = 1</span>
<span>BEGIN</span>                           <span>|</span>
                                <span>|</span> <span>BEGIN</span>
<span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span>   <span>|</span>
<span>COMMIT</span>                          <span>|</span>
                                <span>|</span> <span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span>
                                <span>|</span> <span>COMMIT</span>
</code></pre>
</div>
</div>
<p>The only protection the transaction adds is that (with the default PostgreSQL transaction configuration) it forces the <code>UPDATE</code> in Task 2 to happen after the transaction in Task 1 has committed; when Task 1 runs <code>UPDATE</code> the database automatically places a write lock against the affected records which prevents any other transaction modifying them until Task 1 commits, releasing its locks. It does not prevent the broader race condition caused by concurrent <em>reads</em> that determine the values in the ensuing writes. Further co-ordination is needed to make this work correctly.</p>
<p>Now, this particular example can be implemented quite straightforwardly in SQL, by making the database increment the <code>value</code> rather than copying the record into application memory and incrementing it there:</p>
<div>
<div>
<pre><code><span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>value</span> <span>+</span> <span>1</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'my-counter'</span><span>;</span>
</code></pre>
</div>
</div>
<p>In Rails, this query can be accomplished with this code:</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>Counter</span><span>.</span><span>where</span><span>(</span><span>name: </span><span>"my-counter"</span><span>).</span><span>update_all</span><span>(</span><span>"value = value + 1"</span><span>)</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>However, most application-level problems in the wild aren&rsquo;t so simple, and we usually want to load a record into memory, run some business logic against it, update some values, validate the result, and save it back. This pattern is how most Rails controller actions for updating a record from a form work:</p>
<div>
<div>
<pre><code>  <span>def</span> <span>update</span>
    <span>thing</span> <span>=</span> <span>Thing</span><span>.</span><span>find</span><span>(</span><span>params</span><span>[</span><span>:id</span><span>])</span>
    <span>thing</span><span>.</span><span>update</span><span>(</span><span>params</span><span>[</span><span>:thing</span><span>])</span>
  <span>end</span>
</code></pre>
</div>
</div>
<p>There&rsquo;ll be some additional logic in here, like access control, validation, generating the next page to display, and so on, but this idiom is at the core of how most Rails apps update records, and just like our counter example, it&rsquo;s unsafe unless your server only handles a single request at a time. How do we spot this unsafe pattern? We can see a <code>Location</code> type here: we read a record out of the database, making a copy of it in memory, we change the details of that copy by incorporating some form data into it, and then we write that copy back to the same record in the database, with no other control mechanism around how that update happens.</p>
<p>We&rsquo;ll examine some real-world phenomena that happen with form-based updates in more detail later. First, let&rsquo;s talk about what methods databases like PostgreSQL provide for concurrency control, and how they&rsquo;re exposed in the Rails framework.</p>
<p>One technique you can use is the <code>with_lock</code> method. If we call <code>ctr.update</code> inside a block passed to <code>ctr.with_lock</code>, then the program works correctly and leaves the counter&rsquo;s value at <code>400</code>:</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>ctr</span> <span>=</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
  <span>ctr</span><span>.</span><span>with_lock</span> <span>do</span>
    <span>ctr</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>ctr</span><span>.</span><span>value</span> <span>+</span> <span>1</span><span>)</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p><code>with_lock</code> works using the SQL <code>SELECT FOR UPDATE</code> mechanism. It begins a transaction, and then reloads the <code>ctr</code> object by issuing a <code>SELECT</code> for the record it represents, with <code>FOR UPDATE</code> appended to the query. This makes the database place an exclusive lock against the record, meaning that no concurrent transaction is allowed to read or write it until this transaction completes.</p>
<p>This program generates the following queries. The first <code>SELECT</code> comes from the initial <code>Counter.find_by</code> call. <code>BEGIN</code> and <code>SELECT ... FOR UPDATE</code> are produced by calling <code>ctr.with_lock</code>. The <code>UPDATE</code> query comes from calling <code>ctr.update</code>, and <code>COMMIT</code> happens when the <code>with_lock</code> block exits.</p>
<div>
<div>
<pre><code><span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'my-counter'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>BEGIN</span><span>;</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span> <span>LIMIT</span> <span>1</span> <span>FOR</span> <span>UPDATE</span><span>;</span>
<span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>This makes the program work correctly because it ensures that each logical increment task accesses the counter record sequentially &ndash; the <code>SELECT FOR UPDATE</code> query will be blocked until any current transaction completes, releasing the lock on that record. Above we saw that an <code>UPDATE</code> query cannot be processed until any other transaction currently holding a write lock for the affected record is committed. Adding <code>FOR UPDATE</code> means that the <code>SELECT</code> query also requires a write lock, so it too is blocked until all other transactions are complete. By signalling that we&rsquo;re planning on updating the selected record, we&rsquo;re preemptively locking it to stop anyone else modifying it, so the result of our read remains valid.</p>
<p>The program now produces the right result, but it is wasteful, since we&rsquo;re immediately reloading the record after loading it the first time, and this creates a subtle opportunity to break the program. When we discussed mutexes, we saw that a mutex lock must cover <em>all</em> the reads and writes involved in a task. It is no good to read something, then take a lock before writing it. In this case, it&rsquo;s possible to grab values off the <code>ctr</code> object before it is reloaded, and then use them to increment the counter:</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>ctr</span> <span>=</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
  <span>tmp</span> <span>=</span> <span>ctr</span><span>.</span><span>value</span>

  <span>ctr</span><span>.</span><span>with_lock</span> <span>do</span>
    <span>ctr</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>tmp</span> <span>+</span> <span>1</span><span>)</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p><code>tmp</code> is read before the record is reloaded with <code>FOR UPDATE</code> inside the transaction, so it may be stale: the record may be updated between our first <code>SELECT</code>, and the <code>SELECT FOR UPDATE</code>. That makes the value we stored in <code>tmp</code> invalid, and we end up losing updates just like our original counter example did. In our first <code>with_lock</code> example, we access <code>ctr.value</code> inside the block, once the record has been safely reloaded, whereas in the above example we access it outside the block and its result may not be valid by the time it&rsquo;s used to update the record.</p>
<p>Using values you acquired before locking something is quite a common cause of concurrency bugs; in fact the above is equivalent to how most Rails <code>update</code> actions work. <code>with_lock</code> makes it quite easy to do this without realising, as visually it only appears to wrap the <code>update</code> operation &ndash; it&rsquo;s not obvious that it&rsquo;s implicitly reloading the record so you have the latest version in memory.</p>
<p>This problem can be recognised by the fact that Rails allows the read, write and lock methods to be called on the same type, and in any order; there is nothing to force you to lock the object before interacting with it. As we&rsquo;ve seen, constructing the program so that we know we have a locked object before accessing it will make it safe.</p>
<p>Ultimately <code>with_lock</code> is a convenience method. A method that&rsquo;s less wasteful and harder to accidentally break, is to explicitly start a transaction yourself, and use <code>Counter.lock.find_by</code> to load the record.</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>ActiveRecord</span><span>::</span><span>Base</span><span>.</span><span>transaction</span> <span>do</span>
    <span>ctr</span> <span>=</span> <span>Counter</span><span>.</span><span>lock</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
    <span>ctr</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>ctr</span><span>.</span><span>value</span> <span>+</span> <span>1</span><span>)</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>This does effectively the same thing as the <code>with_lock</code> code, but without the wasted initial <code>SELECT</code>. <code>Counter.lock.find_by</code> performs a <code>SELECT FOR UPDATE</code> query, so we can safely use its result to update the counter&rsquo;s value. This code generates the following queries:</p>
<div>
<div>
<pre><code><span>BEGIN</span><span>;</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'my-counter'</span> <span>LIMIT</span> <span>1</span> <span>FOR</span> <span>UPDATE</span><span>;</span>
<span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>A couple of things to note here: first, the lock created by the <code>FOR UPDATE</code> query is scoped to a transaction, so you can only use this inside a transaction. <code>with_lock</code> takes a block, which means it can start a transaction, run the block, and then commit the transaction for you. <code>Counter.lock</code> doesn&rsquo;t work like this, so you need to supply the transaction for it to run in. This might be annoying but it does make your code visually clearer in terms of signalling the scope of the lock, and which queries are covered by it. It&rsquo;s harder to make the mistake of using stale read results using this construction, because the <code>Counter</code> instance we obtain is in a locked state for its entire lifetime. We&rsquo;ve made the code work like the <code>Mutex</code> type, in that it&rsquo;s only possible to acquire a <code>Counter</code> object once it&rsquo;s safe. <code>Counter.lock.find_by</code> plays the role of <code>Mutex::lock()</code>, giving us an object that&rsquo;s safe to read and write by restricting concurrent access to it.</p>
<p>Second, the use of <code>FOR UPDATE</code> is critical, and merely putting the <code>SELECT</code> and the <code>UPDATE</code> in a transaction together is not sufficient. This implementation, without the <code>lock</code> call, will lose updates:</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>ActiveRecord</span><span>::</span><span>Base</span><span>.</span><span>transaction</span> <span>do</span>
    <span>ctr</span> <span>=</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
    <span>ctr</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>ctr</span><span>.</span><span>value</span> <span>+</span> <span>1</span><span>)</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>This does not lock the counter record while it&rsquo;s in use, which means multiple transactions can run concurrently, for example producing the following sequence of queries:</p>
<div>
<div>
<pre><code><span>Task</span> <span>1</span>                          <span>|</span> <span>Task</span> <span>2</span>
<span>--------------------------------+--------------------------------</span>
<span>BEGIN</span>                           <span>|</span>
                                <span>|</span> <span>BEGIN</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>...</span>      <span>|</span>
    <span>--&gt; value = 1 |</span>
                                <span>|</span> <span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>...</span>
                                <span>|</span>     <span>--&gt; value = 1</span>
<span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span>   <span>|</span>
<span>COMMIT</span>                          <span>|</span>
                                <span>|</span> <span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span>
                                <span>|</span> <span>COMMIT</span>
</code></pre>
</div>
</div>
<p>Task 2 is still allowed to <code>SELECT</code> the record before Task 1 commits, because Task 1&rsquo;s <code>SELECT</code> does not lock the record. Only once Task 1 performs an <code>UPDATE</code> is a lock placed on the record, delaying Task 2&rsquo;s <code>UPDATE</code> until Task 1 commits.</p>
<p>The only way this becomes safe is if you configure PostgreSQL to run transactions using the <em>repeatable read</em> <a href="https://en.wikipedia.org/wiki/Isolation_(database_systems)">isolation level</a>. This causes the database to reject any transaction that attempts to update a record that another concurrent transaction has already written, and requires application-level exception handling:</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>ActiveRecord</span><span>::</span><span>Base</span><span>.</span><span>transaction</span><span>(</span><span>isolation: :repeatable_read</span><span>)</span> <span>do</span>
    <span>ctr</span> <span>=</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
    <span>ctr</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>ctr</span><span>.</span><span>value</span> <span>+</span> <span>1</span><span>)</span>
  <span>end</span>
<span>rescue</span> <span>ActiveRecord</span><span>::</span><span>SerializationFailure</span>
  <span>retry</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>This implementation results in the following queries being sent to PostgreSQL. The <code>UPDATE</code> query will block if another pending transaction has already updated the record, and will return an error once that other transaction commits.</p>
<div>
<div>
<pre><code><span>BEGIN</span> <span>ISOLATION</span> <span>LEVEL</span> <span>REPEATABLE</span> <span>READ</span><span>;</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'my-counter'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>If you run these queries in two separate <code>psql</code> sessions, you&rsquo;ll see the following behaviour:</p>
<div>
<div>
<pre><code><span>Task</span> <span>1</span>                          <span>|</span> <span>Task</span> <span>2</span>
<span>--------------------------------+--------------------------------</span>
<span>BEGIN</span> <span>...</span> <span>REPEATABLE</span> <span>READ</span>       <span>|</span>
                                <span>|</span> <span>BEGIN</span> <span>...</span> <span>REPEATABLE</span> <span>READ</span>
<span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>...</span>      <span>|</span>
    <span>--&gt; value = 1 |</span>
                                <span>|</span> <span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>...</span>
                                <span>|</span>     <span>--&gt; value = 1</span>
<span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span>   <span>|</span>
                                <span>|</span> <span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span>
                                <span>|</span>     <span>--&gt; blocks...</span>
<span>COMMIT</span>                          <span>|</span>
                                <span>|</span>     <span>--&gt; ... and fails</span>
</code></pre>
</div>
</div>
<p>The <code>UPDATE</code> in Task 2 blocks as usual, because Task 1 is holding an exclusive lock on the record after its own <code>UPDATE</code>. The difference in <code>REPEATABLE READ</code> mode is that once Task 1 commits, the <code>UPDATE</code> in Task 2 is rejected, because it conflicts with a write made in Task 1.</p>
<p>Note that you do not need to use <code>Counter.lock</code> to get a <code>SELECT FOR UPDATE</code> here, because <em>repeatable read</em> isolation means the database will automatically prevent concurrent transactions from modifying the same record, without you explicitly locking them. In PostgreSQL the default is the weaker <em>read committed</em> level, which allows the <code>UPDATE</code> in Task 2 above to succeed, even though it&rsquo;s delayed until Task 1 commits.</p>
<p>This has made a very subtle change to the type of the interaction with the counter record: the <code>UPDATE</code> can now fail, which we need to represent in the type somehow. The simplest possible representation that the operation can succeed or fail is to make <code>write(T)</code> return a boolean.</p>
<div>
<div>
<pre><code><span>NonLossy</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>=</span> <span>{</span>
    <span>read</span><span>()</span> <span>-&gt;</span> <span>T</span><span>;</span>
    <span>write</span><span>(</span><span>T</span><span>)</span> <span>-&gt;</span> <span>bool</span><span>;</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>This tiny change to the type means that it&rsquo;s <em>possible</em> (though not guaranteed) for it to prevent lost updates.</p>
<p>It&rsquo;s important to note here that although isolation level terms like <em>read committed</em>, <em>repeatable read</em> and <em>serializable</em> are part of the SQL standard, their definitions are open to interpretation and database implementations may have differing semantics. The above results apply specifically to PostgreSQL; In MySQL, the default isolation level is called &ldquo;repeatable read&rdquo; but it does not prevent the above race condition and both <code>UPDATE</code> queries will succeed. Running this task under the stronger <em>serializable</em> isolation level causes MySQL to <a href="https://en.wikipedia.org/wiki/Deadlock">deadlock</a>. If you want to know how this task works on your database, you should test it yourself. In the case of MySQL, this example is best handled by using the Rails <code>lock</code> method to send a <code>SELECT FOR UPDATE</code> query under any isolation level.</p>
<p>We&rsquo;ve scratched the surface of how databases handle concurrency here, and if you&rsquo;re interested there is substantial documentation of <a href="https://www.postgresql.org/docs/12/mvcc.html">concurrency control in PostgreSQL</a>. There are many different kinds of locks for different use cases, but the key point here is that, to get consistent changes to records, you need to read and write them while holding some kind of transaction-scoped lock that prevents competing tasks from overwriting each other&rsquo;s changes.</p>
<p>This kind of concurrency control is called <em>pessimistic</em>: to prevent conflicting changes, we defensively lock resources before interacting with them. The use of mutexes in the previous article is a form of pessimistic control too. It makes the use of a <code>Location</code> safer by preventing bare read/write access to it, and imposing a control mechanism that only allows one task to use the <code>Location</code> at a time.</p>
<p>There&rsquo;s another kind of control known as <em>optimistic</em>, in which you don&rsquo;t use any locks, you just attempt the desired update, and the update fails if some other task modified the resource at the same time. It&rsquo;s optimistic because you assume there usually won&rsquo;t be any other tasks competing with you for control of a resource, so it would be wasteful to do the work of locking the resource for every single operation.</p>
<p>Optimistic control often takes the form of a <a href="https://en.wikipedia.org/wiki/Compare-and-swap">compare-and-swap</a> (CAS) operation, where a write is made conditional on the current value, or on some secondary value that tracks the resource&rsquo;s <em>current version</em>. In SQL, it might look like a query to set the counter&rsquo;s <code>value</code> if it has not changed since we read it, for example we can set <code>value = 2</code> if and only if the record still has <code>value = 1</code>. This check is implemented as an additional <code>WHERE</code> condition following the <code>id</code> clause.</p>
<div>
<div>
<pre><code><span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span> <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span> <span>AND</span> <span>value</span> <span>=</span> <span>1</span><span>;</span>
</code></pre>
</div>
</div>
<p><code>UPDATE</code> queries return the count of the number of rows affected, so we know if this had any effect and can retry our task if not. The database executes each query atomically, so this doesn&rsquo;t need any transaction around it if we&rsquo;re only performing a single update. In Rails, we can implement this workflow as follows:</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>loop</span> <span>do</span>
    <span>ctr</span>   <span>=</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
    <span>rows</span>  <span>=</span> <span>Counter</span><span>.</span><span>where</span><span>(</span><span>id: </span><span>ctr</span><span>.</span><span>id</span><span>,</span> <span>value: </span><span>ctr</span><span>.</span><span>value</span><span>)</span>
    <span>count</span> <span>=</span> <span>rows</span><span>.</span><span>update_all</span><span>(</span><span>value: </span><span>ctr</span><span>.</span><span>value</span> <span>+</span> <span>1</span><span>)</span>

    <span>break</span> <span>if</span> <span>count</span> <span>&gt;</span> <span>0</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>In a loop, we retrieve the record, and issue an <code>update</code> that&rsquo;s conditional on the <code>value</code> still being the one we just read. If this affects a non-zero number of rows, then we&rsquo;ve successfully updated the record without any other task invalidating our changes, so we can stop retrying.</p>
<p>Rails will do this transparently if you add a <code>lock_version</code> to your table schema:</p>
<div>
<div>
<pre><code><span>ActiveRecord</span><span>::</span><span>Schema</span><span>.</span><span>define</span> <span>do</span>
  <span>create_table</span> <span>:counters</span><span>,</span> <span>force: </span><span>true</span> <span>do</span> <span>|</span><span>t</span><span>|</span>
    <span>t</span><span>.</span><span>integer</span> <span>:lock_version</span><span>,</span> <span>null: </span><span>false</span>
    <span>t</span><span>.</span><span>string</span> <span>:name</span><span>,</span> <span>null: </span><span>false</span>
    <span>t</span><span>.</span><span>integer</span> <span>:value</span><span>,</span> <span>null: </span><span>false</span><span>,</span> <span>default: </span><span>0</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>Now, Rails will take care of the CAS operation, using <code>lock_version</code> as an incrementing number to check if the record has been updated since we read it. The implementation code then has to rescue the exception that Rails throws if the record is stale &ndash; you can retry indefinitely, or some fixed number of times, or just fail immediately.</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>ctr</span> <span>=</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
  <span>ctr</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>ctr</span><span>.</span><span>value</span> <span>+</span> <span>1</span><span>)</span>
<span>rescue</span> <span>ActiveRecord</span><span>::</span><span>StaleObjectError</span>
  <span>retry</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>This produces the correct result and doesn&rsquo;t lose any updates, and does not require the <code>SELECT</code> to be in a transaction with the <code>UPDATE</code>, or to acquire any locks. The transaction in the generated SQL is only there because Rails wraps <em>all</em> updates in transactions; since the <code>SELECT</code> is outside the transaction it&rsquo;s treated as an independent operation by the database.</p>
<div>
<div>
<pre><code><span>SELECT</span> <span>*</span> <span>FROM</span> <span>counters</span> <span>WHERE</span> <span>name</span> <span>=</span> <span>'my-counter'</span> <span>LIMIT</span> <span>1</span><span>;</span>
<span>BEGIN</span><span>;</span>
<span>UPDATE</span> <span>counters</span> <span>SET</span> <span>value</span> <span>=</span> <span>2</span><span>,</span> <span>lock_version</span> <span>=</span> <span>2</span>
    <span>WHERE</span> <span>id</span> <span>=</span> <span>1</span> <span>AND</span> <span>lock_version</span> <span>=</span> <span>1</span><span>;</span>
<span>COMMIT</span><span>;</span>
</code></pre>
</div>
</div>
<p>If another task has updated the record&rsquo;s <code>lock_version</code> since we read <code>lock_version = 1</code> in our <code>SELECT</code>, this will affect zero rows and Rails converts this result into a <code>StaleObjectError</code>.</p>
<p>In terms of our <code>Location</code> type, CAS is a subtle modification to the <code>write(T)</code> operation:</p>
<div>
<div>
<pre><code><span>CompareAndSwap</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>=</span> <span>{</span>
    <span>read</span><span>()</span> <span>-&gt;</span> <span>T</span><span>;</span>
    <span>write</span><span>(</span><span>T</span><span>,</span> <span>T</span><span>)</span> <span>-&gt;</span> <span>bool</span><span>;</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>There&rsquo;s no control mechanism wrapping the resource, so we can still <code>read() -&gt; T</code> from it and get a value whenever we want. The difference is that <code>write(T, T) -&gt; bool</code> takes two values: the new value, <em>and</em> our copy of the last value we read. If that second argument does indeed match the current state, then <code>write()</code> succeeds and returns <code>true</code>, otherwise it returns <code>false</code>. In the <code>Location</code> type, <code>write(T)</code> returns nothing because it always succeeds, so there&rsquo;s no result to report. In <code>CompareAndSwap</code>, it can fail, so it needs to return a representation of that.</p>
<p>This type hints at some additional benefits of optimistic control. It doesn&rsquo;t require locks, and thus allows more work to happen concurrently under the assumption that operations will conflict <em>infrequently</em>. In particular, it allows reads to happen freely; when using pessimistic control, a locking mechanism often blocks writes <em>and</em> reads, so you can&rsquo;t read a resource that somebody else is currently modifying. With optimistic control, any task can read at any time, and writes are controlled by sending additional input, to confirm you&rsquo;ve seen the latest version.</p>
<p>Some databases embrace optimistic concurrency control as part of their core mechanics. In the document database <a href="https://couchdb.apache.org/">CouchDB</a>, all documents have a field called <code>_rev</code> that indicates their current version. Any attempt to update a document needs to supply this value, and the update only succeeds if that version is still the latest one. For example, let&rsquo;s <code>PUT</code> a new <code>counter</code> document whose content is <code>{ "value": 0 }</code> into the database:</p>
<div>
<div>
<pre><code>$ curl -si http://localhost:5984/examples/counter -X PUT -d '{ "value": 0 }'
HTTP/1.1 201 Created
Content-Type: application/json
ETag: "1-7e97409c987eac3a99385a17ad4cbabe"

{"ok":true,"id":"counter","rev":"1-7e97409c987eac3a99385a17ad4cbabe"}
</code></pre>
</div>
</div>
<p>This gives us a successful result, and tells us the <code>_rev</code> that the document has been given &ndash; it&rsquo;s a combination of an incrementing integer and a hash of the document&rsquo;s content.</p>
<p>If we <code>GET</code> this document, we see its current state, including its <code>_rev</code> field.</p>
<div>
<div>
<pre><code>$ curl -si http://localhost:5984/examples/counter -X GET
HTTP/1.1 200 OK
Content-Type: application/json
ETag: "1-7e97409c987eac3a99385a17ad4cbabe"

{"_id":"counter","_rev":"1-7e97409c987eac3a99385a17ad4cbabe","value":0}
</code></pre>
</div>
</div>
<p>If we try to update it using <code>PUT</code>, the request will fail with a <code>409</code> status if we don&rsquo;t supply a <code>_rev</code> value that matches the document&rsquo;s current version.</p>
<div>
<div>
<pre><code>$ curl -si http://localhost:5984/examples/counter -X PUT -d '{ "value": 1 }'
HTTP/1.1 409 Conflict
Content-Type: application/json

{"error":"conflict","reason":"Document update conflict."}
</code></pre>
</div>
</div>
<p>For our write to succeed, we either have to include a <code>_rev</code> field in the replacement document:</p>
<div>
<div>
<pre><code>$ curl -si http://localhost:5984/examples/counter \
      -X PUT -d '{ "value": 1, "_rev": "1-7e97409c987eac3a99385a17ad4cbabe" }'
HTTP/1.1 201 Created
Content-Type: application/json
ETag: "2-610dfd70da9e1879c0addba97d2abbb9"

{"ok":true,"id":"counter","rev":"2-610dfd70da9e1879c0addba97d2abbb9"}
</code></pre>
</div>
</div>
<p>Or, we can pass it in the HTTP headers. HTTP has a built-in optimistic control feature called <a href="https://en.wikipedia.org/wiki/HTTP_ETag">ETags</a>, and if a <code>GET</code> for a resource includes an <code>ETag</code> header in the response, we can make a conditional <code>PUT</code> for the resource using the <code>If-Match</code> header.</p>
<div>
<div>
<pre><code>$ curl -si http://localhost:5984/examples/counter \
      -X PUT -d '{ "value": 2 }' \
      -H 'If-Match: "2-610dfd70da9e1879c0addba97d2abbb9"'
HTTP/1.1 201 Created
Content-Type: application/json
ETag: "3-23fc9ec7ebedcb1eb8d49391bd9a95e8"

{"ok":true,"id":"counter","rev":"3-23fc9ec7ebedcb1eb8d49391bd9a95e8"}
</code></pre>
</div>
</div>
<p>The <code>PUT</code> response includes the new <code>_rev</code> value in the body and in the <code>ETag</code> header, so we immediately know the document&rsquo;s new version and can update it again if required. Making a <code>GET</code> for the document shows that our <code>value</code> updates have indeed been saved.</p>
<div>
<div>
<pre><code>$ curl -si http://localhost:5984/examples/counter -X GET
HTTP/1.1 200 OK
Content-Type: application/json
ETag: "3-23fc9ec7ebedcb1eb8d49391bd9a95e8"

{"_id":"counter","_rev":"3-23fc9ec7ebedcb1eb8d49391bd9a95e8","value":2}
</code></pre>
</div>
</div>
<p>Just like we did in Rails, we can use this to implement a safe counter, by retrying requests when they fail with a conflict. This implementation won&rsquo;t work, because the <code>put</code> call will throw an exception when a conflict happens:</p>
<div>
<div>
<pre><code><span>DB</span> <span>=</span> <span>"http://admin:admin@localhost:5984/examples"</span>
<span>CouchRest</span><span>.</span><span>put</span><span>(</span><span>"</span><span>#{</span><span>DB</span><span>}</span><span>/counter"</span><span>,</span> <span>"value"</span> <span>=&gt;</span> <span>0</span><span>)</span>

<span>def</span> <span>inc_counter</span>
  <span>counter</span> <span>=</span> <span>CouchRest</span><span>.</span><span>get</span><span>(</span><span>"</span><span>#{</span><span>DB</span><span>}</span><span>/counter"</span><span>)</span>
  <span>counter</span><span>[</span><span>"value"</span><span>]</span> <span>+=</span> <span>1</span>
  <span>CouchRest</span><span>.</span><span>put</span><span>(</span><span>"</span><span>#{</span><span>DB</span><span>}</span><span>/counter"</span><span>,</span> <span>counter</span><span>)</span>
<span>end</span>

<span>threads</span> <span>=</span> <span>(</span><span>1</span><span>..</span><span>4</span><span>).</span><span>map</span> <span>do</span>
  <span>Thread</span><span>.</span><span>new</span> <span>do</span>
    <span>100</span><span>.</span><span>times</span> <span>{</span> <span>inc_counter</span> <span>}</span>
  <span>end</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>(The client library I&rsquo;m using is <a href="https://github.com/couchrest/couchrest">CouchRest</a> which provides a thin wrapper over CouchDB&rsquo;s HTTP and JSON interface.) To make this work, all we need to do is retry the operation when it fails:</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>counter</span> <span>=</span> <span>CouchRest</span><span>.</span><span>get</span><span>(</span><span>"</span><span>#{</span><span>DB</span><span>}</span><span>/counter"</span><span>)</span>
  <span>counter</span><span>[</span><span>"value"</span><span>]</span> <span>+=</span> <span>1</span>
  <span>CouchRest</span><span>.</span><span>put</span><span>(</span><span>"</span><span>#{</span><span>DB</span><span>}</span><span>/counter"</span><span>,</span> <span>counter</span><span>)</span>
<span>rescue</span> <span>CouchRest</span><span>::</span><span>Conflict</span>
  <span>retry</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>It&rsquo;s important to note that when using CAS, we do not just retry the write, we retry the whole task. The write fails because we passed in a version argument that&rsquo;s now stale, so we need to read the resource again to get its latest state and try our modification procedure again. This protocol forces the client to prove it&rsquo;s seen the latest version of a resource before being allowed to overwrite it.</p>
<p>Some HTTP-based APIs use this technique, which is effective when it would be inappropriate to let the client hold a long-lived lock over a resource. For example, Dropbox uses <code>ETag</code> and <code>If-Match</code> in its API to prevent two clients causing lost updates, just like CouchDB does, which is especially effective because someone editing a file may have it open for minutes, hours or even days between first loading it and trying to save their changes. However, in my experience it&rsquo;s relatively rare for HTTP services, either public-facing or internal, to provide any form of concurrency control, and such services are vulnerable to all the problems discussed here.</p>
<p>In our <code>CompareAndSwap</code> type above, it&rsquo;s implied that <code>read() -&gt; T</code> and <code>write(T, T) -&gt; bool</code> are independent operations, whereas the example of CouchDB shows that actually one can only write after a successful read for an existing resource. So the type could more accurately be given as:</p>
<div>
<div>
<pre><code><span>CompareAndSwap</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>=</span> <span>{</span>
    <span>read</span><span>()</span> <span>-&gt;</span> <span>Version</span><span>&lt;</span><span>T</span><span>&gt;</span><span>;</span>
<span>}</span>

<span>Version</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>=</span> <span>{</span>
    <span>get</span><span>()</span> <span>-&gt;</span> <span>T</span><span>;</span>
    <span>replace</span><span>(</span><span>T</span><span>)</span> <span>-&gt;</span> <span>bool</span><span>;</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Reading from a <code>CompareAndSwap</code> gives us a <code>Version</code>. Once we have a <code>Version</code>, we can either <code>get()</code> the current value out of it, or <code>replace()</code> it with a new one. <code>replace()</code> implicitly combines the resource&rsquo;s current value (to get its <code>_rev</code> value, for example) with the new value, and sends off an update request that may fail, prompting us to start over with another <code>read()</code> call. One of the nice things about turning protocols into types is it can show us how to design object interfaces so that you can&rsquo;t perform illegal sequences of operations with them; you can&rsquo;t call <code>replace()</code> in this design without calling <code>read()</code> first.</p>
<p>If it&rsquo;s common for operations to conflict, then optimistic concurrency control methods end up spending a lot of time retrying and in that situation it may perform better to preemptively lock resources during modification. Before we move onto to talk about some more application scenarios, I want to mention one final kind of lock that PostgreSQL and other SQL databases offer: <a href="https://www.postgresql.org/docs/12/explicit-locking.html#ADVISORY-LOCKS">advisory locks</a>. These locks aren&rsquo;t directly related to database operations &ndash; the database ignores them when deciding how to schedule or reject transactional reads and writes. Instead they&rsquo;re enforced by your application, so you can protect some of your logic using a mutex that&rsquo;s stored in the database rather than in program memory. That means the mutex protects all your concurrent server processes from each other, not just operations inside a single process.</p>
<p>We can use <code>pg_advisory_lock(id)</code> to implement <code>inc_counter</code>. We&rsquo;ll call it before we try to update the counter record, and PostgreSQL will block if this lock is already held by another connection. When <code>pg_advisory_lock()</code> returns, that means we now hold the lock and can go ahead with our logic, before releasing the lock with <code>pg_advisory_unlock(id)</code>. Rails automatically sets up a separate database connection for each running thread, so this protects threads within the same process from one another, as well as separate processes running the same logic.</p>
<div>
<div>
<pre><code><span>def</span> <span>inc_counter</span>
  <span>Counter</span><span>.</span><span>connection</span><span>.</span><span>execute</span><span>(</span><span>"SELECT pg_advisory_lock(1)"</span><span>)</span>
  <span>ctr</span> <span>=</span> <span>Counter</span><span>.</span><span>find_by</span><span>(</span><span>name: </span><span>"my-counter"</span><span>)</span>
  <span>ctr</span><span>.</span><span>update</span><span>(</span><span>value: </span><span>ctr</span><span>.</span><span>value</span> <span>+</span> <span>1</span><span>)</span>
  <span>Counter</span><span>.</span><span>connection</span><span>.</span><span>execute</span><span>(</span><span>"SELECT pg_advisory_unlock(1)"</span><span>)</span>
<span>end</span>
</code></pre>
</div>
</div>
<p>This lock only has any meaning to the application; the database won&rsquo;t prevent other processes from reading or writing the affected records while this code is running. What it does let you do is selectively co-ordinate around some data, for example to make sure only one background job at a time is allowed to run against a record, while allowing your user-facing application to keep reading and writing it. In this situation, using database-level locks would make data unavailable to end users while your background job is using it, which may be undesirable if you have long-running jobs.</p>
<p>We&rsquo;re now well-armed with a knowledge of the tools databases provide for tackling race conditions, and we can move on to examine data races that occur in web applications &ndash; that&rsquo;s what we&rsquo;ll do in the next instalment.</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via<br>
<br>
June 14, 2022 at 03:23PM
